File: src/main/java/destiny/core/GenderDecorator.java
/**
 * @author smallufo
 * Created on 2008/4/2 at 上午 1:26:30
 */
package destiny.core;

import com.google.common.collect.ImmutableMap;
import destiny.tools.Decorator;
import destiny.tools.LocaleUtils;
import org.jetbrains.annotations.NotNull;

import java.util.Locale;

public class GenderDecorator {
  private final static ImmutableMap<Locale , Decorator<Gender>> implMap = new ImmutableMap.Builder<Locale , Decorator<Gender>>()
    .put(Locale.TAIWAN , new GenderDecoratorChinese())
    .put(Locale.ENGLISH, new GenderDecoratorEnglish())
    .build();

  @NotNull
  public static String getOutputString(Gender gender , Locale locale) {
    return implMap.get(
      LocaleUtils.getBestMatchingLocale(locale, implMap.keySet()).orElse((Locale) implMap.keySet().toArray()[0])
    ).getOutputString(gender);
  }
}

File: src/main/java/destiny/core/GenderDecorator.kt
/**
 * @author smallufo
 * Created on 2008/4/2 at 上午 1:26:30
 */
package destiny.core

import destiny.tools.Decorator
import destiny.tools.getOutputString
import java.util.*

object GenderDecorator {
  private val implMap = mapOf(
    Locale.TAIWAN to GenderDecoratorChinese(),
    Locale.ENGLISH to GenderDecoratorEnglish()
  )

  fun getOutputString(gender: Gender, locale: Locale): String {
    return implMap.getOutputString(gender , locale)
  }
}


class GenderDecoratorChinese : Decorator<Gender> {
  override fun getOutputString(value: Gender): String {
    return if (value === Gender.男) "男" else "女"
  }
}

class GenderDecoratorEnglish : Decorator<Gender> {
  override fun getOutputString(value: Gender): String {
    return if (value === Gender.男) "Male" else "Female"
  }
}
